tripy/tripy/flat_ir/flat_ir.py

```diff
@@ -31,83 +34,160 @@ class FlatIR:
     def __init__(self, shapes: Sequence[ShapeBounds] = None):
         self.inputs: List["FlatIRTensor"] = []
         self.outputs: List["FlatIRTensor"] = []
-        self.ops: List["BaseFlatIROp"] = []
+        self.functions: Dict[str, FlatIRFunction] = {}
+        self.main_flow: List["BaseProducer"] = []
 
         self.shapes = shapes
         self.tensor_map: Dict[str] = {}
 
         self.tensor_replacements: Dict[int, "FlatIRTensor"] = {}
         self.constant_map = {}
 
-    def __str__(self):
-        layer_strs: List[str] = []
-        if len(self.inputs):
-            layer_strs.append("inputs:")
-        for inp in self.inputs:
-            layer_strs.append(f"    {str(inp)}")
-        for op in self.ops:
-            layer_strs.append(str(op))
-        layer_strs.append("outputs:")
-        for out in self.outputs:
-            layer_strs.append(f"    {str(out)}")
-        return "\n".join(layer_strs)
+    def __str__(self) -> str:
+        """Generate a string representation of the FlatIR."""
+        from tripy.flat_ir.ops.base import FlatIRFunction, BaseFlatIROp
+
+        ir_representation = []
+
+        # Print functions
+        for function in self.functions.values():
+            ir_representation.append(str(function))
+            ir_representation.append("")  # Empty line for readability
+
+        # Print main flow
+        ir_representation.append("Main flow:")
```

Comment:
Should we just treat `main` like any other function when we print?

Reply:
That would make the FlatIR print more similar to the MLIR.

Reply:
I think `main` is the only function with nested functions, a `func` can only have base flat ir ops, so treating them as same might be non-trivial.
I am OK to address any concern regarding how we print the IR though.